Mathlib/Analysis/InnerProductSpace/Semisimple.lean

```diff
@@ -0,0 +1,41 @@
+/-
+Copyright (c) 2024 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import Mathlib.Analysis.InnerProductSpace.Projection
+import Mathlib.LinearAlgebra.Semisimple
+
+/-!
+# Semisimple operators on inner product spaces
+
+This file is a place to gather results related to semisimplicity of linear operators on inner
+product spaces.
+
+-/
+
+variable {𝕜 E : Type*} [RCLike 𝕜] [NormedAddCommGroup E] [InnerProductSpace 𝕜 E]
+
+namespace LinearMap.IsSymmetric
+
+variable {T : Module.End 𝕜 E} {p : Submodule 𝕜 E} (hT : T.IsSymmetric)
+
+include hT
+
+/-- The orthogonal complement of an invariant submodule is invariant. -/
+lemma orthogonalComplement_mem_invtSubmodule (hp : p ∈ T.invtSubmodule) :
+    pᗮ ∈ T.invtSubmodule :=
```

Comment:
Can you do the versions of this and the following lemma for `IsStarNormal T` too? Unfortunately, because we don't have an adjoint all the time, here you really will need to restrict to finite dimensional subspaces, and/or to *continuous* linear maps on Hilbert spaces (but for the latter we should probably have a `IsTopologicallySemisimple` predicate, or something to that effect; so if you want to hold off on the continuous ones, that's fine).

The other option would be to phrase it in terms of a commuting `LinearMap.IsAdjointPair`, but even that would require glue to get back to `IsStarNormal` (which I think we don't have currently).

Reply:
For the second day in a row, I've run out of time to implement this suggestion so I'll just say I agree it is worth adding up a version of this for normal operators: they should also be finitely semisimple.

I think it will be a tiny bit of work to prove the result that for a normal operator the orthogonal complement of a _finite-dimensional_ invariant subspace is invariant (without using the spectral theorem) but I hope to get to this at the weekend.

Reply:
I finally had a chance to look into this properly and it seems to me that generalising to normal operators will require a bit of detour. To be precise, the following is the result which I understand you have in mind:
```lean
lemma ContinuousLinearMap.isFinitelySemisimple_of_isStarNormal
    {T : E →L[𝕜] E} [CompleteSpace E] (hT : IsStarNormal T) :
    Module.End.IsFinitelySemisimple (T : Module.End 𝕜 E) :=
  sorry
```
This follows trivially from:
```lean
lemma LinearMap.orthogonalComplement_mem_invtSubmodule [FiniteDimensional 𝕜 E]
    {p : Submodule 𝕜 E} {T : Module.End 𝕜 E} (hT : IsStarNormal T)
    (hp : p ∈ T.invtSubmodule) :
    pᗮ ∈ T.invtSubmodule :=
  sorry
```
This latter result is easy but the best proof I can find (in fact [on Wikipedia](https://en.wikipedia.org/wiki/Normal_operator#Properties_in_finite-dimensional_case)) requires the inner product structure on linear maps. Specifically, we seem to lack the following:
```lean
variable {F : Type*} [NormedAddCommGroup F] [InnerProductSpace 𝕜 F]

open InnerProductSpace InnerProduct

noncomputable instance [FiniteDimensional 𝕜 E] [FiniteDimensional 𝕜 F] :
    Inner 𝕜 (E →L[𝕜] F) where
  inner T S :=
    let _i : CompleteSpace E := FiniteDimensional.complete 𝕜 E
    let _i : CompleteSpace F := FiniteDimensional.complete 𝕜 F
    LinearMap.trace 𝕜 E ((T†).comp S)

instance [FiniteDimensional 𝕜 E] [FiniteDimensional 𝕜 F] :
    InnerProductSpace 𝕜 (E →L[𝕜] F) :=
  sorry
```
This would all be easy to put in place but it's quite a detour.

More generally one should really be assuming compactness instead of finite-dimensionality, i.e.:
```lean
lemma ContinuousLinearMap.orthogonalComplement_mem_invtSubmodule [CompleteSpace E]
    {p : Submodule 𝕜 E} {T : E →L[𝕜] E} (hT : IsStarNormal T) (hT' : IsCompactOperator T)
    (hp : p ∈ Module.End.invtSubmodule T) :
    pᗮ ∈ Module.End.invtSubmodule T :=
  sorry
```
but then this would require some non-trivial work.

So it seems to me there is a nice small project to be done before we are ready for this result. However I don't spend much time in this corner of the library. What do you think?

Reply:
I think there's a more elementary proof IIRC. I'll try to write it down later today, but today is my big teaching day.

Reply:
No rush at all!

Reply:
Unfortunately, the proof I was thinking of establishes a different fact, so it doesn't work here. It would be nice to eventually get the Wold decomposition into Mathlib, but that's for another day, and only tangentially related to this. In any case, let's just skip the normal result for now.